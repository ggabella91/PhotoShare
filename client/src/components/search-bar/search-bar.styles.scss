@mixin hideLabel {
  display: none;
}

@mixin respond-medium {
  @media (max-width: 724px) {
    @content;
  }
}

.search-group {
  position: relative;
  margin-top: 4vh;
  z-index: 20;

  @include respond-medium() {
    transform: translateY(-2vh);
  }

  .search-bar {
    background: none;
    background-color: #7087c8;
    color: white;
    font-size: 14px;
    padding: 0;
    display: block;
    width: 100%;
    min-width: 200px;
    height: 30px;
    border: 1px solid white;
    border-radius: 4px;

    &:focus {
      outline: none;
    }

    &:focus ~ .search-bar-label {
      @include hideLabel();
    }
  }

  input[type='text'] {
    text-indent: 5px;
  }

  .search-bar-label {
    color: white;
    font-size: 14px;
    font-weight: normal;
    position: absolute;
    pointer-events: none;
    left: 5px;
    top: 5px;
    transition: 300ms ease all;

    &.hide {
      @include hideLabel();
    }
  }
}
