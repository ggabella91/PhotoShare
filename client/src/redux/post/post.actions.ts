import { UserInfoAndOtherData } from '../../components/user-info/user-info.component';
import { PostModalDataToFeed } from '../../components/feed-post-container/feed-post-container.component';

import {
  Post,
  PostDataReq,
  PostFileReq,
  Reaction,
  ReactionReq,
  ArchivePostReq,
  DeleteReactionReq,
  PostFile,
  PostError,
  PostActions,
  PostActionTypes,
  PostMetaData,
} from './post.types';

export const createPostStart = (post: FormData): PostActionTypes => ({
  type: PostActions.CREATE_POST_START,
  payload: post,
});

export const createPostSuccess = (post: Post): PostActionTypes => ({
  type: PostActions.CREATE_POST_SUCCESS,
  payload: post,
});

export const createPostFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.CREATE_POST_FAILURE,
  payload: error,
});

export const createPostReactionStart = (
  reactionReq: ReactionReq
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_START,
  payload: reactionReq,
});

export const createPostReactionSuccess = (
  reaction: Reaction
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_SUCCESS,
  payload: reaction,
});

export const createPostReactionFailure = (
  error: PostError
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_FAILURE,
  payload: error,
});

export const getPostReactionsStart = (postId: string): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_START,
  payload: postId,
});

export const getPostReactionsSuccess = (
  reactions: Reaction[]
): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_SUCCESS,
  payload: reactions,
});

export const getPostReactionsFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_FAILURE,
  payload: error,
});

export const clearPostStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_STATUSES,
  payload: null,
});

export const updateProfilePhotoStart = (photo: FormData): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_START,
  payload: photo,
});

export const updateProfilePhotoSuccess = (photo: Post): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_SUCCESS,
  payload: photo,
});

export const updateProfilePhotoFailure = (
  error: PostError
): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_FAILURE,
  payload: error,
});

export const clearProfilePhotoStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_PROFILE_PHOTO_STATUSES,
  payload: null,
});

export const getPostDataStart = (
  postDataReq: PostDataReq
): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_START,
  payload: postDataReq,
});

export const getPostDataSuccess = (postDataArray: Post[]): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_SUCCESS,
  payload: postDataArray,
});

export const addPostDataToFeedArray = (
  postDataArray: Post[]
): PostActionTypes => ({
  type: PostActions.ADD_POST_DATA_TO_FEED_ARRAY,
  payload: postDataArray,
});

export const getPostDataFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_FAILURE,
  payload: error,
});

export const getPostFileStart = (fileReq: PostFileReq): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_START,
  payload: fileReq,
});

export const getPostFileSuccess = (file: PostFile): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_SUCCESS,
  payload: file,
});

export const getPostFileFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_FAILURE,
  payload: error,
});

export const getProfilePhotoFileSuccess = (
  file: PostFile
): PostActionTypes => ({
  type: PostActions.GET_PROFILE_PHOTO_FILE_SUCCESS,
  payload: file,
});

export const archivePostStart = (
  archiveReq: ArchivePostReq
): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_START,
  payload: archiveReq,
});

export const archivePostSuccess = (message: string): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_SUCCESS,
  payload: message,
});

export const archivePostFailure = (error: Error): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_FAILURE,
  payload: error,
});

export const deleteReactionStart = (
  deleteReactionReq: DeleteReactionReq
): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_START,
  payload: deleteReactionReq,
});

export const deleteReactionSuccess = (message: string): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_SUCCESS,
  payload: message,
});

export const deleteReactionFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_FAILURE,
  payload: error,
});

export const getOtherUserProfilePhotoFileSuccess = (
  file: string
): PostActionTypes => ({
  type: PostActions.GET_OTHER_USER_PROFILE_PHOTO_FILE_SUCCESS,
  payload: file,
});

export const getUserPhotoForFollowArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_FOLLOW_ARRAY_SUCCESS,
  payload: postFile,
});

export const getUserPhotoForSuggestionArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_SUGGESTION_ARRAY_SUCCESS,
  payload: postFile,
});

export const getUserPhotoForReactorArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_REACTOR_ARRAY_SUCCESS,
  payload: postFile,
});

export const clearArchivePostStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_ARCHIVE_POST_STATUSES,
  payload: null,
});

export const clearFollowPhotoFileArray = (): PostActionTypes => ({
  type: PostActions.CLEAR_FOLLOW_PHOTO_FILE_ARRAY,
  payload: null,
});

export const clearSuggestionPhotoFileArray = (): PostActionTypes => ({
  type: PostActions.CLEAR_SUGGESTION_PHOTO_FILE_ARRAY,
  payload: null,
});

export const clearPostFilesAndData = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_FILES_AND_DATA,
  payload: null,
});

export const clearPostState = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_STATE,
  payload: null,
});

export const setCommentToDelete = (
  deleteReactionReq: DeleteReactionReq
): PostActionTypes => ({
  type: PostActions.SET_COMMENT_TO_DELETE,
  payload: deleteReactionReq,
});

export const setShowCommentOptionsModal = (
  showCommentOptionsModal: boolean
): PostActionTypes => ({
  type: PostActions.SET_SHOW_COMMENT_OPTIONS_MODAL,
  payload: showCommentOptionsModal,
});

export const clearPostReactions = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_REACTIONS,
  payload: null,
});

export const setPostMetaDataForUser = (
  postMetaData: PostMetaData
): PostActionTypes => ({
  type: PostActions.SET_POST_META_DATA_FOR_USER,
  payload: postMetaData,
});

export const setPostLikingUsersArray = (
  postLikingUsersArray: UserInfoAndOtherData[]
): PostActionTypes => ({
  type: PostActions.SET_POST_LIKING_USERS_ARRAY,
  payload: postLikingUsersArray,
});

export const setShowPostLikingUsersModal = (
  showPostLikingUsersModal: boolean
): PostActionTypes => ({
  type: PostActions.SET_SHOW_POST_LIKING_USERS_MODAL,
  payload: showPostLikingUsersModal,
});

export const setFeedPagePostModalData = (
  postModalDataToFeed: PostModalDataToFeed
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_MODAL_DATA,
  payload: postModalDataToFeed,
});

export const setFeedPagePostModalShow = (
  postModalDataShow: boolean
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_MODAL_SHOW,
  payload: postModalDataShow,
});

export const setFeedPagePostOptionsModalShow = (
  postOptionsModalDataShow: boolean
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_OPTIONS_MODAL_SHOW,
  payload: postOptionsModalDataShow,
});

export const setClearFeedPagePostModalState = (
  clearFeedPagePostModalState: boolean
): PostActionTypes => ({
  type: PostActions.SET_CLEAR_FEED_PAGE_POST_MODAL_STATE,
  payload: clearFeedPagePostModalState,
});

// Actions for feed-post-container

export const getFeedPostFileSuccess = (file: PostFile): PostActionTypes => ({
  type: PostActions.GET_FEED_POST_FILE_SUCCESS,
  payload: file,
});

export const getFeedPostReactionsSuccess = (
  reactions: Reaction[]
): PostActionTypes => ({
  type: PostActions.GET_FEED_POST_REACTIONS_SUCCESS,
  payload: reactions,
});

export const getUserPhotoForFeedReactorArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_FEED_REACTOR_ARRAY_SUCCESS,
  payload: postFile,
});
