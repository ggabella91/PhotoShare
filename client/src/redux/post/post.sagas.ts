import { takeLatest, takeEvery, put, all, call } from 'redux-saga/effects';
import { SagaIterator } from '@redux-saga/core';
import { ActionPattern, Saga } from '@redux-saga/types';

import {
  Post,
  ReactionReq,
  ReactionRequestType,
  GetPostReactionsReq,
  DataRequestType,
  FileRequestType,
  PostDataReq,
  PostFileReq,
  ArchivePostReq,
  DeleteReactionReq,
  PostActions,
  UserType,
  EditPostDetailsReq,
  Reaction,
  SinglePostDataReq,
  PostError,
  PostsWithHashtagReq,
} from './post.types';

import {
  createPostSuccess,
  createPostFailure,
  createPostReactionSuccess,
  createPostReactionFailure,
  updateProfilePhotoSuccess,
  updateProfilePhotoFailure,
  addToPostDataArray,
  addPostDataToFeedArray,
  getPostDataFailure,
  getPostReactionsSuccess,
  getFeedPostReactionsSuccess,
  getPostReactionsFailure,
  getPostFileSuccess,
  getProfilePhotoFileSuccess,
  getPostFileFailure,
  archivePostSuccess,
  archivePostFailure,
  deleteReactionSuccess,
  deleteReactionFailure,
  getOtherUserProfilePhotoFileSuccess,
  getUserPhotoForFollowArraySuccess,
  getUserPhotoForSuggestionArraySuccess,
  getUserPhotoForReactorArraySuccess,
  setPostMetaDataForUser,
  getFeedPostFileSuccess,
  getUserPhotoForFeedReactorArraySuccess,
  editPostDetailsSuccess,
  editPostDetailsFailure,
  getSinglePostDataSuccess,
  getSinglePostDataFailure,
  setPostMetaDataForHashtag,
} from './post.actions';

import axios from 'axios';

export function* createPost({ payload: post }: { payload: FormData }): any {
  try {
    const { data } = yield axios.post('/api/posts/new', post);

    yield put(createPostSuccess(data));
  } catch (err) {
    yield put(createPostFailure(err as PostError));
  }
}

export function* createPostReaction({
  payload: reactionReq,
}: {
  payload: ReactionReq;
}): any {
  try {
    const { data }: { data: Reaction } = yield axios.post(
      '/api/reactions/new',
      reactionReq
    );

    yield put(
      createPostReactionSuccess({
        reactionId: data.id,
        likedPost: data.likedPost,
        message: '',
        postId: data.postId,
      })
    );
  } catch (err) {
    yield put(createPostReactionFailure(err as PostError));
  }
}

export function* updateProfilePhoto({
  payload: photo,
}: {
  payload: FormData;
}): any {
  try {
    const { data } = yield axios.post('/api/posts/profilePhoto', photo);

    yield put(updateProfilePhotoSuccess(data));
  } catch (err) {
    yield put(updateProfilePhotoFailure(err as PostError));
  }
}

export function* getPostData({
  payload: { userId, dataReqType, pageToShow, limit },
}: {
  payload: PostDataReq;
}): any {
  try {
    const { data }: { data: { posts: Post[]; queryLength?: number } } =
      yield axios.get(
        `/api/posts/data?userId=${userId}&pageToShow=${pageToShow}&limit=${limit}`
      );

    if (dataReqType === DataRequestType.single) {
      if (data.queryLength) {
        yield all([
          put(addToPostDataArray(data.posts)),
          put(
            setPostMetaDataForUser({
              queryLength: data.queryLength,
              userId: data.posts[0].userId,
            })
          ),
        ]);
      } else {
        yield put(addToPostDataArray(data.posts));
      }
    } else if (dataReqType === DataRequestType.feed) {
      if (data.queryLength) {
        yield all([
          put(addPostDataToFeedArray(data.posts)),
          put(
            setPostMetaDataForUser({
              queryLength: data.queryLength,
              userId: data.posts[0].userId,
            })
          ),
        ]);
      } else {
        yield put(addPostDataToFeedArray(data.posts));
      }
    }
  } catch (err) {
    yield put(getPostDataFailure(err as PostError));
  }
}

export function* getPostsWithHashtag({
  payload: { hashtag, pageToShow, limit },
}: {
  payload: PostsWithHashtagReq;
}): any {
  try {
    const { data }: { data: { posts: Post[]; queryLength?: number } } =
      yield axios.get(
        `/api/posts/hashtags/${hashtag}?pageToShow=${pageToShow}&limit=${limit}`
      );

    if (data.queryLength) {
      yield all([
        put(addToPostDataArray(data.posts)),
        put(
          setPostMetaDataForHashtag({
            queryLength: data.queryLength,
            hashtag,
          })
        ),
      ]);
    } else {
      yield put(addToPostDataArray(data.posts));
    }
  } catch (err) {
    yield put(getPostDataFailure(err as PostError));
  }
}

export function* getPostReactions({
  payload: { postId, reactionReqType },
}: {
  payload: GetPostReactionsReq;
}): any {
  try {
    const { data } = yield axios.get(`/api/reactions/${postId}`);

    if (reactionReqType === ReactionRequestType.singlePost) {
      yield put(getPostReactionsSuccess(data));
    } else if (reactionReqType === ReactionRequestType.feedPost) {
      yield put(getFeedPostReactionsSuccess(data));
    }
  } catch (err) {
    yield put(getPostReactionsFailure(err as PostError));
  }
}

export function* getPostFile({
  payload: { s3Key, bucket, user, fileRequestType },
}: {
  payload: PostFileReq;
}): any {
  try {
    const { data } = yield axios.get(
      `/api/posts/files?s3Key=${s3Key}&bucket=${bucket}`
    );

    if (bucket === 'photo-share-app' || bucket === 'photo-share-app-dev') {
      if (fileRequestType === FileRequestType.singlePost) {
        yield put(getPostFileSuccess({ s3Key, fileString: data }));
      } else if (fileRequestType === FileRequestType.feedPost) {
        yield put(getFeedPostFileSuccess({ s3Key, fileString: data }));
      }
    } else if (
      bucket === 'photo-share-app-profile-photos' ||
      bucket === 'photo-share-app-profile-photos-dev'
    ) {
      if (user === UserType.self) {
        yield put(getProfilePhotoFileSuccess({ s3Key, fileString: data }));
      } else if (user === UserType.other) {
        yield put(
          getOtherUserProfilePhotoFileSuccess({ s3Key, fileString: data })
        );
      } else if (user === UserType.followArray) {
        yield put(
          getUserPhotoForFollowArraySuccess({ s3Key, fileString: data })
        );
      } else if (user === UserType.suggestionArray) {
        yield put(
          getUserPhotoForSuggestionArraySuccess({ s3Key, fileString: data })
        );
      } else if (
        fileRequestType === FileRequestType.singlePost &&
        user === UserType.postReactorsArray
      ) {
        yield put(
          getUserPhotoForReactorArraySuccess({ s3Key, fileString: data })
        );
      } else if (
        fileRequestType === FileRequestType.feedPost &&
        user === UserType.postReactorsArray
      ) {
        yield put(
          getUserPhotoForFeedReactorArraySuccess({ s3Key, fileString: data })
        );
      }
    }
  } catch (err) {
    yield put(getPostFileFailure(err as PostError));
  }
}

export function* archivePost({
  payload: { postId, s3Key },
}: {
  payload: ArchivePostReq;
}): any {
  try {
    const { data } = yield axios.delete(`/api/posts/${postId}`, {
      data: {
        s3Key,
      },
    });

    yield put(archivePostSuccess(data.message));
  } catch (err) {
    yield put(archivePostFailure(err as PostError));
  }
}

export function* deleteReaction({
  payload: deleteReactionReq,
}: {
  payload: DeleteReactionReq;
}): any {
  try {
    const { data } = yield axios.delete(`/api/reactions`, {
      data: deleteReactionReq,
    });

    if (deleteReactionReq.isLikeRemoval) {
      yield put(
        deleteReactionSuccess({
          reactionId: data.reactionId,
          message: 'Like removed successfully!',
          postId: deleteReactionReq.postId,
        })
      );
    } else {
      yield put(
        deleteReactionSuccess({
          reactionId: data.reactionId,
          message: 'Comment removed successfully!',
          postId: deleteReactionReq.postId,
        })
      );
    }
  } catch (err) {
    yield put(deleteReactionFailure(err as PostError));
  }
}

export function* editPostDetails({
  payload: { postId, caption, location },
}: {
  payload: EditPostDetailsReq;
}): any {
  try {
    const { data }: { data: Post } = yield axios.patch(`/api/posts/${postId}`, {
      caption,
      location,
    });

    yield put(editPostDetailsSuccess(data));
  } catch (err) {
    yield put(editPostDetailsFailure(err as PostError));
  }
}

export function* getSinglePostData({
  payload: { postId },
}: {
  payload: SinglePostDataReq;
}): any {
  try {
    const { data }: { data: Post } = yield axios.get(
      `/api/posts/data/${postId}`
    );

    yield put(getSinglePostDataSuccess(data));
  } catch (err) {
    yield put(getSinglePostDataFailure(err as PostError));
  }
}

export function* onCreatePostStart(): SagaIterator {
  yield takeEvery<ActionPattern, Saga>(
    PostActions.CREATE_POST_START,
    createPost
  );
}

export function* onCreatePostReactionStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    PostActions.CREATE_POST_REACTION_START,
    createPostReaction
  );
}

export function* onUpdateProfilePhotoStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    PostActions.UPDATE_PROFILE_PHOTO_START,
    updateProfilePhoto
  );
}

export function* onGetPostDataStart(): SagaIterator {
  yield takeEvery<ActionPattern, Saga>(
    PostActions.GET_POST_DATA_START,
    getPostData
  );
}

export function* onGetPostsWithHashtagStart(): SagaIterator {
  yield takeEvery<ActionPattern, Saga>(
    PostActions.GET_POSTS_WITH_HASHTAG_START,
    getPostsWithHashtag
  );
}

export function* onGetPostReactionsStart(): SagaIterator {
  yield takeEvery<ActionPattern, Saga>(
    PostActions.GET_POST_REACTIONS_START,
    getPostReactions
  );
}

export function* onGetPostFileStart(): SagaIterator {
  yield takeEvery<ActionPattern, Saga>(
    PostActions.GET_POST_FILE_START,
    getPostFile
  );
}

export function* onArchivePostStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    PostActions.ARCHIVE_POST_START,
    archivePost
  );
}

export function* onDeleteReactionStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    PostActions.DELETE_REACTION_START,
    deleteReaction
  );
}

export function* onEditPostDetailsStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    PostActions.EDIT_POST_DETAILS_START,
    editPostDetails
  );
}

export function* onGetSinglePostDataStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    PostActions.GET_SINGLE_POST_DATA_START,
    getSinglePostData
  );
}

export function* postSagas(): SagaIterator {
  yield all([
    call(onCreatePostStart),
    call(onCreatePostReactionStart),
    call(onUpdateProfilePhotoStart),
    call(onGetPostDataStart),
    call(onGetPostsWithHashtagStart),
    call(onGetPostReactionsStart),
    call(onGetPostFileStart),
    call(onArchivePostStart),
    call(onDeleteReactionStart),
    call(onEditPostDetailsStart),
    call(onGetSinglePostDataStart),
  ]);
}
