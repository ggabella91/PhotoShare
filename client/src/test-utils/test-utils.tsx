import { FC, ReactElement } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render, RenderOptions } from '@testing-library/react';
import { store } from '../redux/store';

interface WrapperProps {
  wrapperProps?: { route: string; location: string };
}

type CustomRenderOptions = RenderOptions & WrapperProps;

const WithReduxProvider: FC<WrapperProps> = ({ children, wrapperProps }) => {
  return (
    <Provider store={store}>
      <BrowserRouter>
        <Routes>
          <Route path={wrapperProps?.route || '/'} element={children} />
        </Routes>
      </BrowserRouter>
    </Provider>
  );
};

const renderWithReduxAndRouter = (
  ui: ReactElement,
  options?: Omit<CustomRenderOptions, 'wrapper'>
) => {
  window.history.pushState({}, '', options?.wrapperProps?.location || '/');
  const customContainer = document.createElement('div');
  customContainer.id = 'root';

  return render(ui, {
    wrapper: (props) => (
      <WithReduxProvider {...props} wrapperProps={options?.wrapperProps} />
    ),
    container: document.body.appendChild(customContainer),
    ...options,
  });
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { renderWithReduxAndRouter as render };
