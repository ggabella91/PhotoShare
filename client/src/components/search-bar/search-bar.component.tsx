import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { createStructuredSelector } from 'reselect';
import { AppState } from '../../redux/root-reducer';

import { User, Error } from '../../redux/user/user.types';
import {
  selectUserSuggestions,
  selectUserSuggestionsError,
} from '../../redux/user/user.selectors';
import { getUserSuggestionsStart } from '../../redux/user/user.actions';

import { PostFile, PostFileReq } from '../../redux/post/post.types';
import {} from '../../redux/post/post.selectors';
import { getPostFileStart } from '../../redux/post/post.actions';

import UserSuggestions from '../user-suggestions/user-suggestions.component';

import './search-bar.styles.scss';

export interface SearchBarProps {
  userSuggestions: User[];
  userSuggestionsError: Error | null;
  getUserSuggestionsStart: typeof getUserSuggestionsStart;
  getPostFileStart: typeof getPostFileStart;
}

export interface UserSuggestionsData {
  profilePhotoFileString: string;
  username: string;
  name: string;
}

const SearchBar: React.FC<SearchBarProps> = ({
  getUserSuggestionsStart,
  userSuggestions,
  userSuggestionsError,
}) => {
  const [searchString, setSearchString] = useState('');
  const [userSuggestionsArray, setUserSuggestionsArray] = useState<
    UserSuggestionsData[]
  >([]);
  const [showUserSuggestions, setShowUserSuggestions] = useState(false);

  const handleSearchStringChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const { value } = event.target;

    setSearchString(value);
  };

  useEffect(() => {
    if (searchString.length >= 3) {
      getUserSuggestionsStart(searchString);
    } else {
      setShowUserSuggestions(false);
    }
  }, [searchString]);

  useEffect(() => {
    if (userSuggestions) {
      const suggested: UserSuggestionsData[] = userSuggestions.map(
        (el: User) => {
          return {
            name: el.name,
            username: el.username,
            profilePhotoFileString: '',
          };
        }
      );

      setUserSuggestionsArray(suggested);
    }
  }, [userSuggestions]);

  useEffect(() => {
    if (userSuggestionsArray.length) {
      setShowUserSuggestions(true);
    } else {
      setShowUserSuggestions(false);
    }
  }, [userSuggestionsArray]);

  return (
    <form>
      <div className='search-group'>
        <label
          className={`${searchString.length ? 'hide' : ''} search-bar-label`}
        >
          Search
        </label>
        <input
          onChange={handleSearchStringChange}
          className='search-bar'
          name='search'
          type='text'
          value={searchString}
        />
        <UserSuggestions
          userSuggestionsArray={userSuggestionsArray}
          show={showUserSuggestions}
        />
      </div>
    </form>
  );
};

interface LinkStateProps {
  userSuggestions: User[];
  userSuggestionsError: Error | null;
}

const mapStateToProps = createStructuredSelector<AppState, LinkStateProps>({
  userSuggestions: selectUserSuggestions,
  userSuggestionsError: selectUserSuggestionsError,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  getUserSuggestionsStart: (match: string) =>
    dispatch(getUserSuggestionsStart(match)),
  getPostFileStart: (fileReq: PostFileReq) =>
    dispatch(getPostFileStart(fileReq)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchBar);
