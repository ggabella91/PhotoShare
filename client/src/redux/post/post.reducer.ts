import { PostActions, PostActionTypes, PostState } from './post.types';

const INITIAL_STATE: PostState = {
  postData: null,
  postFiles: [],
  getPostDataError: null,
  getPostDataConfirm: null,
  getPostFileError: null,
  getPostFileConfirm: null,
  postError: null,
  postConfirm: null,
  profilePhotoFile: null,
  profilePhotoError: null,
  profilePhotoConfirm: null,
};

const postReducer = (state = INITIAL_STATE, action: PostActionTypes) => {
  switch (action.type) {
    case PostActions.UPDATE_PROFILE_PHOTO_SUCCESS:
      return {
        ...state,
        profilePhotoError: null,
        profilePhotoConfirm: 'Profile photo updated!',
      };
    case PostActions.CREATE_POST_SUCCESS:
      return {
        ...state,
        postError: null,
        postConfirm: 'Post uploaded!',
      };
    case PostActions.GET_POST_DATA_SUCCESS:
      return {
        ...state,
        postData: action.payload,
        getPostDataError: null,
        getPostDataConfirm: 'Post data fetched!',
      };
    case PostActions.GET_POST_FILE_SUCCESS:
      return {
        ...state,
        postFiles: [...state.postFiles, action.payload],
        getPostFileError: null,
        getPostFileConfirm: 'Post file fetched!',
      };
    case PostActions.GET_PROFILE_PHOTO_FILE_SUCCESS:
      return {
        ...state,
        profilePhotoFile: action.payload,
        getPostFileError: null,
        getPostFileConfirm: 'Profile photo file fetched!',
      };
    case PostActions.UPDATE_PROFILE_PHOTO_FAILURE:
      return {
        ...state,
        profilePhotoError: action.payload,
        profilePhotoConfirm: null,
      };
    case PostActions.CREATE_POST_FAILURE:
      return {
        ...state,
        postError: action.payload,
        postConfirm: null,
      };
    case PostActions.GET_POST_DATA_FAILURE:
      return {
        ...state,
        getPostDataError: action.payload,
        getPostDataConfirm: null,
      };
    case PostActions.GET_POST_FILE_FAILURE:
      return {
        ...state,
        getPostFileError: action.payload,
        getPostFileConfirm: null,
      };
    default:
      return state;
  }
};

export default postReducer;
