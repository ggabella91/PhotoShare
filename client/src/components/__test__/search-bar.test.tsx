import { render, screen, userEvent } from '../../test-utils/test-utils';

import { SearchBar } from '../search-bar/search-bar.component';

import { User } from '../../redux/user/user.types';
import '../../redux/post/post.actions';
import { PostFile } from '../../redux/post/post.types';

describe('renders a search-bar component', () => {
  const setup = () => {
    const testUser = {} as User;
    const testUserArray = [] as User[];
    const testPostFileArray = [] as PostFile[];

    const getUserSuggestionsStart = jest.fn();
    const getPostFileStart = jest.fn();
    const clearUserSuggestions = jest.fn();
    const clearSuggestionPhotoFileArray = jest.fn();

    render(
      <SearchBar
        key={0}
        currentUser={testUser}
        getUserSuggestionsStart={getUserSuggestionsStart}
        userSuggestions={testUserArray}
        userSuggestionsError={null}
        getPostFileStart={getPostFileStart}
        userSuggestionProfilePhotoFiles={testPostFileArray}
        userSuggestionsConfirm='confirm'
        userSuggestionProfilePhotoConfirm='confirm'
        clearUserSuggestions={clearUserSuggestions}
        clearSuggestionPhotoFileArray={clearSuggestionPhotoFileArray}
      />
    );

    return {
      getUserSuggestionsStart,
      getPostFileStart,
      clearUserSuggestions,
      clearSuggestionPhotoFileArray,
    };
  };

  it('renders a search bar component', () => {
    setup();

    const searchBar = screen.getByTestId('search-bar');

    expect(searchBar).toBeInTheDocument();
  });

  it('typing three or more characters in the search bar results in the user suggestion container component being rendered with no matches found', () => {
    setup();

    const searchBarInput = screen.getByRole('textbox');

    userEvent.type(searchBarInput, 'test');

    const noMatchesSpan = screen.getByText(/No matches found/i);

    expect(noMatchesSpan).toBeInTheDocument();
  });
});
