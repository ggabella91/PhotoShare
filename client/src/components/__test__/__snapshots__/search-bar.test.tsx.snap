// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a search-bar component 1`] = `
<form>
  <div
    className="search-group"
  >
    <label
      className=" search-bar-label"
    >
      Search
    </label>
    <input
      className="search-bar"
      name="search"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value=""
    />
  </div>
</form>
`;
